runtime/compiler/env/DependencyTable.cpp

```diff
@@ -0,0 +1,169 @@
+   {
+   auto ramClass = (J9Class *)clazz;
+
+   uintptr_t classOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(clazz, &classOffset))
+      return;
+
+   // We only need to check if clazz matches its cached version on load; on
+   // initialization, it will be in the _offsetMap if it did match.
+   if (isClassLoad && !_sharedCache->classMatchesCachedVersion(ramClass, NULL))
+      return;
+
+   OMR::CriticalSection cs(_tableMonitor);
+
+   auto entry = getOffsetEntry(classOffset, isClassLoad);
+   if (isClassLoad)
+      entry->_loadedClasses.insert(ramClass);
+   }
+
+OffsetEntry *
+TR_AOTDependencyTable::getOffsetEntry(uintptr_t offset, bool create)
+   {
+   auto it = _offsetMap.find(offset);
+   if (it != _offsetMap.end())
+      return &it->second;
+
+   if (create)
+      {
+      PersistentUnorderedSet<J9Class *> loadedClasses(PersistentUnorderedSet<J9Class *>::allocator_type(TR::Compiler->persistentAllocator()));
+      return &(*_offsetMap.insert(it, {offset, {loadedClasses}})).second;
+      }
+
+   return NULL;
+   }
+
+void
+TR_AOTDependencyTable::invalidateUnloadedClass(TR_OpaqueClassBlock *clazz)
+   {
+   if (!_sharedCache)
+      return;
+
+   uintptr_t classOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(clazz, &classOffset))
+      return;
+
+
+   OMR::CriticalSection cs(_tableMonitor);
+
+   auto it = _offsetMap.find(classOffset);
+   auto entry = getOffsetEntry(classOffset, false);
+   if (entry)
+      entry->_loadedClasses.erase((J9Class *)clazz);
+   }
+
+// If an entry exists for a class, remove it. Otherwise, if we should
+// revalidate, add an entry if the class has a valid chain.
+void
+TR_AOTDependencyTable::recheckSubclass(J9Class *ramClass, uintptr_t offset, bool shouldRevalidate)
+   {
+   auto it = _offsetMap.find(offset);
+   if (it != _offsetMap.end())
+      {
+      auto c_it = it->second._loadedClasses.find(ramClass);
+      if (c_it != it->second._loadedClasses.end())
+         it->second._loadedClasses.erase(c_it);
+      else if (_sharedCache->classMatchesCachedVersion(ramClass, NULL))
+         it->second._loadedClasses.insert(c_it, ramClass);
+      }
+   else if (shouldRevalidate && _sharedCache->classMatchesCachedVersion(ramClass, NULL))
+      {
+      PersistentUnorderedSet<J9Class *> loadedClasses(PersistentUnorderedSet<J9Class *>::allocator_type(TR::Compiler->persistentAllocator()));
+      loadedClasses.insert(ramClass);
+      _offsetMap.insert(it, {offset, {loadedClasses}});
+      }
+   }
+
+// In a class redefinition event, an old class is replaced by a fresh class. If
+// the ROM class offset changed as a result, it and all its subclasses that
+// formerly had valid chains will now be guaranteed not to match, so the entries
+// for these must be removed. If the new offset is valid, any class that didn't
+// have an entry should be rechecked.
```

Comment:
This comment says that if the ROM class offset changes, the entries must be removed. However, further down you have
```
   if (oldClassOffset != freshClassOffset)
      return;
```
which means that if the offset changed, you don't do anything, which contradicts the comment.

Reply:
Also, if the offset is different, then you definitely need to update the table. Otherwise, the table will have an entry with the `oldClassOffset` as the key but the `freshClass` as the value (because the `freshClass` is redefined in place). You may not need to recheck the subclass though.

Reply:
Typo introduced when I rewrote a bit of this. It should be `==`. If the offsets are the same then the validation status remains unchanged.